specs/ContextMenuRequested.md

```diff
@@ -0,0 +1,477 @@
+
+        /// Get the Shortcut for the Context Menu Item
+        [propget] HRESULT Shortcut([out, retval] LPWSTR* value);
+
+        /// Returns the menu item kind
+        [propget] HRESULT Kind([out, retval] COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND* value);
+
+        /// Returns if the menu item is enabled
+        [propget] HRESULT IsEnabled([out, retval] BOOL* value);
+
+        /// Returns if the menu item is checked, used if the kind is checkbox or radio
+        [propget] HRESULT IsChecked([out, retval] BOOL* value);
+
+        /// Returns the list of Children menu items if the kind is Submenu
+        [propget] HRESULT Children([out, retval] ICoreWebView2ContextMenuItemCollection** value);
+
+    }
+
+    /// Collection of ContextMenuItem objects 
+    [uuid(f562a2f5-c415-45cf-b909-d4b7c1e276d3), object, pointer_default(unique)]
+    interface ICoreWebView2ContextMenuItemCollection : IUnknown {
+        /// The number of context menu items contained in the ICoreWebView2ContextMenuItemCollection
+        [propget] HRESULT Count([out, retval] UINT* value);
+
+        /// Gets the context menu item at the given index.
+        HRESULT GetValueAtIndex([in] UINT index,
+            [out, retval] ICoreWebView2ContextMenuItem** value);
+
+        // Removes ContextMenuItem at the specified index
+        HRESULT RemoveValueAtIndex([in] UINT index);
+
+        /// Will insert the new ContextMenuItem at the index specified
+        HRESULT AddValueAtIndex([in] UINT index, [in] ICoreWebView2ContextMenuItem* value);
+    }
+
+    [uuid(76eceacb-0462-4d94-ac83-423a6793775e), object, pointer_default(unique)]
+    interface ICoreWebView2_4 : ICoreWebView2_3
+    {
+        /// Add an event handler for the ContextMenuRequested event.
+        /// ContextMenuRequested event is raied when the user right clicks in the webview
+        ///
+        /// The host can use their own UI to create their own context menu using
+        /// the data provided in the API or can add to / remove from the default
+        /// context menu. If the host doesn't handle the event, Webview will
+        /// display the original context menu.
+
+        HRESULT add_ContextMenuRequested(
+            [in] ICoreWebView2ContextMenuRequestedEventHandler* eventHandler,
+            [out] EventRegistrationToken* token);
+
+        /// Remove an event handler previously added with add_ContextMenuRequested.
+        HRESULT remove_ContextMenuRequested(
+            [in] EventRegistrationToken token);
+    }
+
+    /// A continuation of the ICoreWebView2Environment interface.
+    [uuid(04d4fe1d-ab87-42fb-a898-da241d35b63c), object, pointer_default(unique)]
+    interface ICoreWebView2Environment : IUnknown {
+        /// Create a ContextMenuItem object used for developers to insert new items 
+        /// into the default browser context menu, Descriptor will be CUSTOM
+        HRESULT CreateContextMenuItem(
```

Comment:
You have new properties on your context menu item like enabled, checked and so on that the end developer cannot modify. Do you want to have them as parameters in CreateContextMenuItem or make them settable properties on the context menu item interface?

Reply:
I think it makes more sense to include them as parameters in the CreateContextMenuItem since they shouldn't be settable once the context menu gets shown. Checked shouldn't be settable because it's state will change once the end user clicks on it. Enabled shouldn't be settable either because if the end developer wants to disable an item, they might as well remove it completely. Checked, enabled etc are just for displaying the custom context menu correctly and should be parameters for CreateContextMenuItem.